File: leetcode/problems/language-specific/go/74.search-a-2-d-matrix.go
/*
 * @lc app=leetcode id=74 lang=golang
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) <= 0 || len(matrix[0]) <= 0 {
		return false
	}

	m, n := len(matrix), len(matrix[0])
	left, right := 0, m*n-1
	for left <= right {
		mid := left + (right-left)/2
		row, col := valToPos(mid, n)
		if matrix[row][col] == target {
			return true
		} else if matrix[row][col] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return false
}

// valToPos returns the corresponding position (row and col) in 2D array from 1D array
func valToPos(val, n int) (int, int) {
	return int(val / n), val % n
}

// @lc code=end

File: leetcode/problems/problem-specific/74.search-a-2-d-matrix/74.search-a-2-d-matrix.py
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#

# @lc code=start

# * Binary Search Solution | O(log(mn)) Time | O(1) Space
# * m -> The number of rows in the matrix | n -> The number of columns in the matrix


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROWS, COLS = len(matrix), len(matrix[0])

        def get_val_to_pos(val):
            """Returns the corresponding position (row and col) in 2D array from 1D array."""
            return (val // COLS), (val % COLS)

        left, right = 0, (ROWS * COLS) - 1
        while left <= right:
            mid = left + (right - left) // 2
            row, col = get_val_to_pos(mid)
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                left = mid + 1
            else:
                right = mid - 1

        return False


# @lc code=end
